### ISSUE TITLE
Freeze while passing NaN to unwrap_phase

### ISSUE THREAD

**Issue by ldes89150** [scikit-image/scikit-image#3831]
## Description
I found in some situation the program will freeze after passing the array contains NaN values. Here is the minimum example I found that can reproduce the error. I'd like to know if there is a way to deal with missing data.
Thanks!

## Way to reproduce
```python
import numpy as np                                                      
from skimage.restoration import unwrap_phase   

x = np.linspace(1,100)   
x[1] = np.nan    
xx ,yy = np.meshgrid(x,x)   
unwrap_phase(xx)
```

## Version information
```
3.7.1 (default, Dec 14 2018, 19:28:38) 
[GCC 7.3.0]
Linux-4.9.0-8-amd64-x86_64-with-debian-9.8
scikit-image version: 0.14.2
numpy version: 1.15.4
```

**Comment by jni** [comment]
@ldes89150 thanks for the report! The code should never freeze, that's definitely a bug. However, to indicate masked values, you should use a NumPy masked array. The second part of the gallery example shows how to use this:

http://scikit-image.org/docs/dev/auto_examples/filters/plot_phase_unwrap.html

@scikit-image/core longer term, perhaps we should add a `mask` argument to the function that has the same semantics as the rest of the library?

**Comment by jni** [comment]
@ldes89150 thanks for the report! The code should never freeze, that's definitely a bug. However, to indicate masked values, you should use a NumPy masked array. The second part of the gallery example shows how to use this:

http://scikit-image.org/docs/dev/auto_examples/filters/plot_phase_unwrap.html

@scikit-image/core longer term, perhaps we should add a `mask` argument to the function that has the same semantics as the rest of the library?

**Comment by stefanv** [comment]
@ldes89150 bug confirmed.
@jni agreed.

**Comment by hmaarrfk** [comment]
Quickly looking at the code, it seems there are some floating point comparisons that aren't NaN safe in the C code.

Not sure the best way to deal with this that won't completely kill performance.

**Comment by stefanv** [comment]
On Thu, 11 Apr 2019 16:42:53 -0700, Mark Harfouche wrote:
> Not sure the best way to deal with this that won't completely kill
> performance.

It's okay to have weird results after NaNs are introduced; it just
shouldn't freeze.

**Comment by hmaarrfk** [comment]
Agreed. I think it is the floating point comparison that is tripping off the algorithm

**Comment by jni** [comment]
```c
typedef enum {
  yes,
  no
} yes_no;

yes_no find_pivot(EDGE *left, EDGE *right, double *pivot_ptr) {
```

Really, [bools](https://www.geeksforgeeks.org/bool-in-c/) were not sufficiently expressive here? 🤦‍♂️ 

@hmaarrfk a hang suggests a while loop is involved rather than a for-loop. I notice two whiles later on in that function...

@stefanv perhaps `np.nan_to_num` would be useful here? =)

**Comment by hmaarrfk** [comment]
We should look at how other NaN safe functions are implemented

**Comment by ldes89150** [comment]
@jni Thanks for the suggestion of using masked arrays!

**Comment by plu2k** [comment]
Masking arrays unfortunately does not work for me. I have a 3000 x 3000 float array with NaNs masked out using MaskedArrays. Testing numpy.isnan(arr).any() returns False. 
Maybe it is worth mentioning that the NaNs cover about 25% of the "outer" regions of the array including complete rows / columns. The rest is well connected, though. I did check value ranges.

**Comment by jni** [comment]
@plu2k are you able to share your masked array with us for testing?

**Comment by plu2k** [comment]
This is strange. Although using masked arrays works in some cases, when trying to corner the problem I found this minimal example:

```python
import numpy as np
from skimage.restoration import unwrap_phase
wrapped = np.empty((100, 100), dtype=np.float64)
wrapped[:] = np.linspace(0, 4*np.pi, 100) % (2*np.pi)
wrapped[10, 10] = np.nan  #  <-- comment out to get working code
masked = np.ma.masked_array(wrapped, mask=np.isnan(wrapped))
print(np.isnan(masked).any())
unwrapped = unwrap_phase(masked)
```
which causes a freeze (skimage 0.15.0, numpy 1.16.2).

**Comment by plu2k** [comment]
... whereas this works:
```python
import numpy as np
from skimage.restoration import unwrap_phase
wrapped = np.empty((100, 100), dtype=np.float64)
wrapped[:] = np.linspace(0, 4*np.pi, 100) % (2*np.pi)
wrapped[10, 10] = np.nan  #  <-- comment out to get working code
masked = np.ma.masked_array(wrapped, mask=np.isnan(wrapped))
print(np.isnan(masked).any())
unwrapped = unwrap_phase(2*masked)  # multiply by factor 20
```
O.o

**Comment by grlee77** [comment]
> This is strange. Although using masked arrays works in some cases, when trying to corner the problem I found this minimal 

Thanks, @plu2k, I can confirm that this is still the case on v0.19.2

**Comment by lagru** [comment]
See https://github.com/scikit-image/scikit-image/pull/7176 for an attempted fix.

**Comment by stefanv** [comment]
Tagging @matthew-brett who recently worked on unwrap code.
